
# Project Objectives
---

<div style="text-align: justify"> 
Provider Fraud is one of the biggest problems facing Medicare. According to the government, the total Medicare spending increased exponentially due to frauds in Medicare claims. Healthcare fraud is an organized crime which involves peers of providers, physicians, beneficiaries acting together to make fraud claims. Rigorous analysis of Medicare data has yielded many physicians who indulge in fraud. They adopt ways in which an ambiguous diagnosis code is used to adopt costliest procedures and drugs. Insurance companies are the most vulnerable institutions impacted due to these bad practices. Due to this reason, insurance companies increased their insurance premiums and as result healthcare is becoming costly matter day by day.</div>

Healthcare fraud and abuse take many forms. Some of the most common types of frauds by providers are:

- Billing for services that were not provided.

- Duplicate submission of a claim for the same service.

- Misrepresenting the service provided.

- Charging for a more complex or expensive service than was actually provided.

- Billing for a covered service when the service actually provided was not covered.

    


# Problem Statement
---
<div style="text-align: justify">The goal of this project is to " predict the potentially fraudulent providers " based on the claims filed by them. Along with this, we will also discover important variables helpful in detecting the behaviour of potentially fraud providers. Further, we will study fraudulent patterns in the provider's claims to understand the future behaviour of providers.</div>


# Introduction to the Dataset
---
<div style="text-align: justify">For the purpose of this project, we are considering Inpatient claims, Outpatient claims and Beneficiary details of each provider. Lets s see their details :</div>
    
**1. Inpatient Data**: This data provides insights about the claims filed for those patients who are admitted in the hospitals. It also provides additional details like their admission and discharge dates and admit d diagnosis code.

**2. Outpatient Data**: This data provides details about the claims filed for those patients who visit hospitals and not admitted in it.
    
**3. Beneficiary Details Data**: This data contains beneficiary KYC details like health conditions,region they belong to etc.

In [1]:
# Importing necessary libraries
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import zipfile

# 1. Importing Data

In [2]:
provider_data=pd.read_csv(r"Data\provider_fraud_detection_data.csv")
provider_data

Provider PotentialFraud
0     PRV51001             No
1     PRV51003            Yes
2     PRV51004             No
3     PRV51005            Yes
4     PRV51007             No
...        ...            ...
5405  PRV57759             No
5406  PRV57760             No
5407  PRV57761             No
5408  PRV57762             No
5409  PRV57763             No

[5410 rows x 2 columns]

In [3]:
data_list=['outpatientdata','inpatientdata','beneficiarydata']
dataframes=[]
for i in data_list:
    zipfile_path=f'Data/{i}.zip'
    zf=zipfile.ZipFile(zipfile_path)
    for file in zf.namelist():
        print (f"files in {zipfile_path} : {file}")
        df=pd.read_csv(zf.open(file))
        dataframes.append(df)

files in Data/outpatientdata.zip : Train_Outpatientdata-1542865627584.csv
files in Data/inpatientdata.zip : Train_Inpatientdata-1542865627584.csv
files in Data/beneficiarydata.zip : Train_Beneficiarydata-1542865627584.csv


In [4]:
outpatient_data=dataframes[0]
inpatient_data=dataframes[1]
beneficiary_data=dataframes[2]

# 2. Understand the Data

### a. Provider Data

In [32]:
provider_data.head()

Provider PotentialFraud
0  PRV51001             No
1  PRV51003            Yes
2  PRV51004             No
3  PRV51005            Yes
4  PRV51007             No

In [33]:
provider_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5410 entries, 0 to 5409
Data columns (total 2 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Provider        5410 non-null   object
 1   PotentialFraud  5410 non-null   object
dtypes: object(2)
memory usage: 84.7+ KB


In [34]:
provider_data.nunique()

Provider          5410
PotentialFraud       2
dtype: int64

In [35]:
x=provider_data.groupby(['PotentialFraud']).count()
x

Provider
PotentialFraud          
No                  4904
Yes                  506

#### There is 506 potential fraud providers. There is a possibility that they are either filing correct or incorrect bills.

### b. Outpatient Data

In [36]:
outpatient_data.head()

BeneID    ClaimID ClaimStartDt  ClaimEndDt  Provider  \
0  BENE11002  CLM624349   2009-10-11  2009-10-11  PRV56011   
1  BENE11003  CLM189947   2009-02-12  2009-02-12  PRV57610   
2  BENE11003  CLM438021   2009-06-27  2009-06-27  PRV57595   
3  BENE11004  CLM121801   2009-01-06  2009-01-06  PRV56011   
4  BENE11004  CLM150998   2009-01-22  2009-01-22  PRV56011   

   InscClaimAmtReimbursed AttendingPhysician OperatingPhysician  \
0                      30          PHY326117                NaN   
1                      80          PHY362868                NaN   
2                      10          PHY328821                NaN   
3                      40          PHY334319                NaN   
4                     200          PHY403831                NaN   

  OtherPhysician ClmDiagnosisCode_1  ... ClmDiagnosisCode_9  \
0            NaN              78943  ...                NaN   
1            NaN               6115  ...                NaN   
2            NaN               2723  ...                NaN   
3            NaN              71988  ...                NaN   
4            NaN              82382  ...                NaN   

  ClmDiagnosisCode_10 ClmProcedureCode_1 ClmProcedureCode_2  \
0                 NaN                NaN                NaN   
1                 NaN                NaN                NaN   
2                 NaN                NaN                NaN   
3                 NaN                NaN                NaN   
4                 NaN                NaN                NaN   

  ClmProcedureCode_3 ClmProcedureCode_4 ClmProcedureCode_5 ClmProcedureCode_6  \
0                NaN                NaN                NaN                NaN   
1                NaN                NaN                NaN                NaN   
2                NaN                NaN                NaN                NaN   
3                NaN                NaN                NaN                NaN   
4                NaN                NaN                NaN                NaN   

  DeductibleAmtPaid  ClmAdmitDiagnosisCode  
0                 0                  56409  
1                 0                  79380  
2                 0                    NaN  
3                 0                    NaN  
4                 0                  71947  

[5 rows x 27 columns]

In [51]:
print(f"Columns: {len(outpatient_data.columns)} \n   Rows: {len(outpatient_data)}")

Columns: 27 
   Rows: 517737


In [52]:
outpatient_data_info=pd.DataFrame(data=outpatient_data.columns, 
                                 columns=['Column Name'])

In [53]:
outpatient_data_info['Data Type']=outpatient_data.dtypes.values
outpatient_data_info['No.of Unique Values']=outpatient_data.nunique().values
outpatient_data_info['Count of Missing Nos']=outpatient_data.isna().sum().values
outpatient_data_info['Percent of Missing Nos']=(round(outpatient_data.isna().sum()/len(outpatient_data)*100)).values

In [54]:
outpatient_data_info

Column Name Data Type  No.of Unique Values  \
0                   BeneID    object               133980   
1                  ClaimID    object               517737   
2             ClaimStartDt    object                  385   
3               ClaimEndDt    object                  366   
4                 Provider    object                 5012   
5   InscClaimAmtReimbursed     int64                  342   
6       AttendingPhysician    object                74109   
7       OperatingPhysician    object                28532   
8           OtherPhysician    object                44388   
9       ClmDiagnosisCode_1    object                10354   
10      ClmDiagnosisCode_2    object                 5056   
11      ClmDiagnosisCode_3    object                 4448   
12      ClmDiagnosisCode_4    object                 3925   
13      ClmDiagnosisCode_5    object                 3412   
14      ClmDiagnosisCode_6    object                 2968   
15      ClmDiagnosisCode_7    object                 2635   
16      ClmDiagnosisCode_8    object                 2260   
17      ClmDiagnosisCode_9    object                 1894   
18     ClmDiagnosisCode_10    object                  495   
19      ClmProcedureCode_1   float64                   80   
20      ClmProcedureCode_2   float64                   22   
21      ClmProcedureCode_3   float64                    4   
22      ClmProcedureCode_4   float64                    2   
23      ClmProcedureCode_5   float64                    0   
24      ClmProcedureCode_6   float64                    0   
25       DeductibleAmtPaid     int64                   16   
26   ClmAdmitDiagnosisCode    object                 3715   

    Count of Missing Nos  Percent of Missing Nos  
0                      0                     0.0  
1                      0                     0.0  
2                      0                     0.0  
3                      0                     0.0  
4                      0                     0.0  
5                      0                     0.0  
6                   1396                     0.0  
7                 427120                    82.0  
8                 322691                    62.0  
9                  10453                     2.0  
10                195380                    38.0  
11                314480                    61.0  
12                392141                    76.0  
13                443393                    86.0  
14                468981                    91.0  
15                484776                    94.0  
16                494825                    96.0  
17                502899                    97.0  
18                516654                   100.0  
19                517575                   100.0  
20                517701                   100.0  
21                517733                   100.0  
22                517735                   100.0  
23                517737                   100.0  
24                517737                   100.0  
25                     0                     0.0  
26                412312                    80.0

### c. Inpatient Data

In [55]:
inpatient_data.head()

BeneID   ClaimID ClaimStartDt  ClaimEndDt  Provider  \
0  BENE11001  CLM46614   2009-04-12  2009-04-18  PRV55912   
1  BENE11001  CLM66048   2009-08-31  2009-09-02  PRV55907   
2  BENE11001  CLM68358   2009-09-17  2009-09-20  PRV56046   
3  BENE11011  CLM38412   2009-02-14  2009-02-22  PRV52405   
4  BENE11014  CLM63689   2009-08-13  2009-08-30  PRV56614   

   InscClaimAmtReimbursed AttendingPhysician OperatingPhysician  \
0                   26000          PHY390922                NaN   
1                    5000          PHY318495          PHY318495   
2                    5000          PHY372395                NaN   
3                    5000          PHY369659          PHY392961   
4                   10000          PHY379376          PHY398258   

  OtherPhysician AdmissionDt  ... ClmDiagnosisCode_7  ClmDiagnosisCode_8  \
0            NaN  2009-04-12  ...               2724               19889   
1            NaN  2009-08-31  ...                NaN                 NaN   
2      PHY324689  2009-09-17  ...                NaN                 NaN   
3      PHY349768  2009-02-14  ...              25062               40390   
4            NaN  2009-08-13  ...               5119               29620   

  ClmDiagnosisCode_9 ClmDiagnosisCode_10 ClmProcedureCode_1  \
0               5849                 NaN                NaN   
1                NaN                 NaN             7092.0   
2                NaN                 NaN                NaN   
3               4019                 NaN              331.0   
4              20300                 NaN             3893.0   

  ClmProcedureCode_2 ClmProcedureCode_3 ClmProcedureCode_4 ClmProcedureCode_5  \
0                NaN                NaN                NaN                NaN   
1                NaN                NaN                NaN                NaN   
2                NaN                NaN                NaN                NaN   
3                NaN                NaN                NaN                NaN   
4                NaN                NaN                NaN                NaN   

  ClmProcedureCode_6  
0                NaN  
1                NaN  
2                NaN  
3                NaN  
4                NaN  

[5 rows x 30 columns]

In [57]:
print(f"Columns: {len(inpatient_data.columns)} \n   Rows: {len(inpatient_data)}")

Columns: 30 
   Rows: 40474


In [58]:
inpatient_data_info=pd.DataFrame(data=inpatient_data.columns, 
                                 columns=['Column Name'])

In [59]:
inpatient_data_info['Data Type']= inpatient_data.dtypes.values
inpatient_data_info['No.of Unique Values']= inpatient_data.nunique().values
inpatient_data_info['Count of Missing Nos']= inpatient_data.isna().sum().values
inpatient_data_info['Percent of Missing Nos']=(round(inpatient_data.isna().sum()/len(inpatient_data)*100)).values

In [60]:
inpatient_data_info

Column Name Data Type  No.of Unique Values  \
0                   BeneID    object                31289   
1                  ClaimID    object                40474   
2             ClaimStartDt    object                  398   
3               ClaimEndDt    object                  365   
4                 Provider    object                 2092   
5   InscClaimAmtReimbursed     int64                  146   
6       AttendingPhysician    object                11604   
7       OperatingPhysician    object                 8287   
8           OtherPhysician    object                 2877   
9              AdmissionDt    object                  398   
10   ClmAdmitDiagnosisCode    object                 1928   
11       DeductibleAmtPaid   float64                    1   
12             DischargeDt    object                  365   
13      DiagnosisGroupCode    object                  736   
14      ClmDiagnosisCode_1    object                 2254   
15      ClmDiagnosisCode_2    object                 2439   
16      ClmDiagnosisCode_3    object                 2427   
17      ClmDiagnosisCode_4    object                 2441   
18      ClmDiagnosisCode_5    object                 2374   
19      ClmDiagnosisCode_6    object                 2358   
20      ClmDiagnosisCode_7    object                 2310   
21      ClmDiagnosisCode_8    object                 2243   
22      ClmDiagnosisCode_9    object                 2094   
23     ClmDiagnosisCode_10    object                  952   
24      ClmProcedureCode_1   float64                 1117   
25      ClmProcedureCode_2   float64                  297   
26      ClmProcedureCode_3   float64                  154   
27      ClmProcedureCode_4   float64                   48   
28      ClmProcedureCode_5   float64                    6   
29      ClmProcedureCode_6   float64                    0   

    Count of Missing Nos  Percent of Missing Nos  
0                      0                     0.0  
1                      0                     0.0  
2                      0                     0.0  
3                      0                     0.0  
4                      0                     0.0  
5                      0                     0.0  
6                    112                     0.0  
7                  16644                    41.0  
8                  35784                    88.0  
9                      0                     0.0  
10                     0                     0.0  
11                   899                     2.0  
12                     0                     0.0  
13                     0                     0.0  
14                     0                     0.0  
15                   226                     1.0  
16                   676                     2.0  
17                  1534                     4.0  
18                  2894                     7.0  
19                  4838                    12.0  
20                  7258                    18.0  
21                  9942                    25.0  
22                 13497                    33.0  
23                 36547                    90.0  
24                 17326                    43.0  
25                 35020                    87.0  
26                 39509                    98.0  
27                 40358                   100.0  
28                 40465                   100.0  
29                 40474                   100.0

### d. Finding columns shared by inpatient and outpatient data 

In [67]:
# Finding Common columns
a = inpatient_data.columns.intersection(outpatient_data.columns)
print(len(a))


27


In [69]:
inpatient_data.shape

(40474, 30)

In [79]:
print("The columns which are only in the inpatient data:")
for i in inpatient_data.columns:
    #print (i)
    if i not in a:
        print(f"  {i}")

The columns which are only in the inpatient data:
  AdmissionDt
  DischargeDt
  DiagnosisGroupCode


####  There is claim start date and claim end date in the inpatient data. We need to check whether the 
    1. "ClaimStartDt=AdmissionDt" and 
    2. "ClaimEndDt=DischargeDt"

In [91]:
(inpatient_data['ClaimStartDt']==inpatient_data['AdmissionDt'])

0        True
1        True
2        True
3        True
4        True
         ... 
40469    True
40470    True
40471    True
40472    True
40473    True
Length: 40474, dtype: bool

In [92]:
(inpatient_data['ClaimStartDt']==inpatient_data['AdmissionDt']).sum()

40442

In [102]:
x=pd.DataFrame(inpatient_data.loc[inpatient_data['ClaimStartDt']!=inpatient_data['AdmissionDt']]['ClaimStartDt'])

In [105]:
x['AdmissionDt']=inpatient_data.loc[inpatient_data['ClaimStartDt']!=inpatient_data['AdmissionDt']]['AdmissionDt']

In [106]:
x

ClaimStartDt AdmissionDt
2163    2009-02-10  2009-02-08
4422    2009-10-02  2009-09-29
5400    2009-09-12  2009-09-09
5647    2009-03-07  2009-03-02
5869    2009-07-10  2009-07-06
7828    2009-04-20  2009-04-16
7869    2009-11-01  2009-10-24
9051    2009-11-26  2009-11-18
9215    2009-01-11  2009-01-06
11103   2009-03-28  2009-03-26
11570   2009-11-14  2009-11-13
12418   2009-07-16  2009-07-14
12763   2009-03-18  2009-03-04
13660   2009-09-13  2009-09-11
13974   2009-08-31  2009-08-29
16043   2009-08-19  2009-08-18
17023   2009-01-16  2008-12-30
17581   2009-01-26  2008-12-29
18270   2009-10-21  2009-10-18
22977   2009-08-22  2009-08-21
24250   2009-04-24  2009-04-15
25600   2009-02-26  2009-01-30
27400   2009-02-15  2009-02-04
27576   2009-04-07  2009-04-06
28093   2009-05-22  2009-05-20
28246   2009-04-21  2009-03-26
28355   2009-11-17  2009-11-14
28398   2009-03-18  2009-03-16
29157   2009-03-02  2009-02-21
31734   2009-08-19  2009-07-24
35222   2009-11-19  2009-11-15
39683   2009-04-10  2009-04-09

### e. Beneficiary Details Data

In [19]:
beneficiary_data.head()

BeneID         DOB  DOD  Gender  Race RenalDiseaseIndicator  State  \
0  BENE11001  1943-01-01  NaN       1     1                     0     39   
1  BENE11002  1936-09-01  NaN       2     1                     0     39   
2  BENE11003  1936-08-01  NaN       1     1                     0     52   
3  BENE11004  1922-07-01  NaN       1     1                     0     39   
4  BENE11005  1935-09-01  NaN       1     1                     0     24   

   County  NoOfMonths_PartACov  NoOfMonths_PartBCov  ...  \
0     230                   12                   12  ...   
1     280                   12                   12  ...   
2     590                   12                   12  ...   
3     270                   12                   12  ...   
4     680                   12                   12  ...   

   ChronicCond_Depression  ChronicCond_Diabetes  ChronicCond_IschemicHeart  \
0                       1                     1                          1   
1                       2                     2                          2   
2                       2                     2                          1   
3                       2                     1                          1   
4                       2                     1                          2   

   ChronicCond_Osteoporasis  ChronicCond_rheumatoidarthritis  \
0                         2                                1   
1                         2                                2   
2                         2                                2   
3                         1                                1   
4                         2                                2   

   ChronicCond_stroke  IPAnnualReimbursementAmt  IPAnnualDeductibleAmt  \
0                   1                     36000                   3204   
1                   2                         0                      0   
2                   2                         0                      0   
3                   2                         0                      0   
4                   2                         0                      0   

   OPAnnualReimbursementAmt  OPAnnualDeductibleAmt  
0                        60                     70  
1                        30                     50  
2                        90                     40  
3                      1810                    760  
4                      1790                   1200  

[5 rows x 25 columns]

In [20]:
f"The inpatient data has {len(beneficiary_data.columns)} columns and {len(beneficiary_data)} rows"

'The inpatient data has 25 columns and 138556 rows'

In [21]:
beneficiary_data_info=pd.DataFrame(data=beneficiary_data.columns, 
                                 columns=['Column Name'])

In [22]:
beneficiary_data_info['Data Type']= beneficiary_data.dtypes.values
beneficiary_data_info['No.of Unique Values']= beneficiary_data.nunique().values
beneficiary_data_info['Count of Missing Nos']= beneficiary_data.isna().sum().values
beneficiary_data_info['Percent of Missing Nos']=(round(beneficiary_data.isna().sum()/len(beneficiary_data)*100)).values

In [23]:
beneficiary_data_info

Column Name Data Type  No.of Unique Values  \
0                            BeneID    object               138556   
1                               DOB    object                  900   
2                               DOD    object                   11   
3                            Gender     int64                    2   
4                              Race     int64                    4   
5             RenalDiseaseIndicator    object                    2   
6                             State     int64                   52   
7                            County     int64                  314   
8               NoOfMonths_PartACov     int64                   13   
9               NoOfMonths_PartBCov     int64                   13   
10            ChronicCond_Alzheimer     int64                    2   
11         ChronicCond_Heartfailure     int64                    2   
12        ChronicCond_KidneyDisease     int64                    2   
13               ChronicCond_Cancer     int64                    2   
14       ChronicCond_ObstrPulmonary     int64                    2   
15           ChronicCond_Depression     int64                    2   
16             ChronicCond_Diabetes     int64                    2   
17        ChronicCond_IschemicHeart     int64                    2   
18         ChronicCond_Osteoporasis     int64                    2   
19  ChronicCond_rheumatoidarthritis     int64                    2   
20               ChronicCond_stroke     int64                    2   
21         IPAnnualReimbursementAmt     int64                 3004   
22            IPAnnualDeductibleAmt     int64                  147   
23         OPAnnualReimbursementAmt     int64                 2078   
24            OPAnnualDeductibleAmt     int64                  789   

    Count of Missing Nos  Percent of Missing Nos  
0                      0                     0.0  
1                      0                     0.0  
2                 137135                    99.0  
3                      0                     0.0  
4                      0                     0.0  
5                      0                     0.0  
6                      0                     0.0  
7                      0                     0.0  
8                      0                     0.0  
9                      0                     0.0  
10                     0                     0.0  
11                     0                     0.0  
12                     0                     0.0  
13                     0                     0.0  
14                     0                     0.0  
15                     0                     0.0  
16                     0                     0.0  
17                     0                     0.0  
18                     0                     0.0  
19                     0                     0.0  
20                     0                     0.0  
21                     0                     0.0  
22                     0                     0.0  
23                     0                     0.0  
24                     0                     0.0

# 3. Data Preprocessing

### Merging Inpatient and Outpatient Data Columns

In [24]:
len(set(outpatient_data.columns).intersection(set(inpatient_data.columns)))

27

In [25]:
len(inpatient_data.columns)

30

In [26]:
in_out_patient_data=outpatient_data.append(inpatient_data, ignore_index = True)

In [27]:
in_out_patient_data.shape

(558211, 30)

In [30]:
outpatient_data.shape[0]+inpatient_data.shape[0]

558211

### Joining provider details (Fraudulent or not)

In [31]:
in_out_patient_data=pd.merge(in_out_patient_data,provider_data,on='Provider')
in_out_patient_data

BeneID    ClaimID ClaimStartDt  ClaimEndDt  Provider  \
0        BENE11002  CLM624349   2009-10-11  2009-10-11  PRV56011   
1        BENE11004  CLM121801   2009-01-06  2009-01-06  PRV56011   
2        BENE11004  CLM150998   2009-01-22  2009-01-22  PRV56011   
3        BENE11004  CLM173224   2009-02-03  2009-02-03  PRV56011   
4        BENE11004  CLM224741   2009-03-03  2009-03-03  PRV56011   
...            ...        ...          ...         ...       ...   
558206  BENE149280   CLM68702   2009-09-19  2009-09-24  PRV54054   
558207  BENE151892   CLM31767   2008-12-31  2009-01-02  PRV51136   
558208  BENE154375   CLM39644   2009-02-22  2009-02-25  PRV51977   
558209  BENE156181   CLM57017   2009-06-24  2009-07-01  PRV55706   
558210  BENE159166   CLM38985   2009-02-18  2009-02-25  PRV51812   

        InscClaimAmtReimbursed AttendingPhysician OperatingPhysician  \
0                           30          PHY326117                NaN   
1                           40          PHY334319                NaN   
2                          200          PHY403831                NaN   
3                           20          PHY339887                NaN   
4                           40          PHY345721                NaN   
...                        ...                ...                ...   
558206                    5000          PHY405930                NaN   
558207                   20000          PHY361876          PHY361876   
558208                    8000          PHY323189          PHY349527   
558209                   12000          PHY319644          PHY319644   
558210                    6000          PHY430902          PHY430902   

       OtherPhysician ClmDiagnosisCode_1  ... ClmProcedureCode_3  \
0                 NaN              78943  ...                NaN   
1                 NaN              71988  ...                NaN   
2                 NaN              82382  ...                NaN   
3                 NaN              20381  ...                NaN   
4                 NaN              V6546  ...                NaN   
...               ...                ...  ...                ...   
558206            NaN                486  ...                NaN   
558207            NaN              44481  ...                NaN   
558208            NaN              44024  ...                NaN   
558209      PHY403799              78079  ...                NaN   
558210            NaN               1629  ...                NaN   

       ClmProcedureCode_4 ClmProcedureCode_5 ClmProcedureCode_6  \
0                     NaN                NaN                NaN   
1                     NaN                NaN                NaN   
2                     NaN                NaN                NaN   
3                     NaN                NaN                NaN   
4                     NaN                NaN                NaN   
...                   ...                ...                ...   
558206                NaN                NaN                NaN   
558207                NaN                NaN                NaN   
558208                NaN                NaN                NaN   
558209                NaN                NaN                NaN   
558210                NaN                NaN                NaN   

       DeductibleAmtPaid ClmAdmitDiagnosisCode AdmissionDt DischargeDt  \
0                    0.0                 56409         NaN         NaN   
1                    0.0                   NaN         NaN         NaN   
2                    0.0                 71947         NaN         NaN   
3                    0.0                   NaN         NaN         NaN   
4                    0.0                   NaN         NaN         NaN   
...                  ...                   ...         ...         ...   
558206            1068.0                   486  2009-09-19  2009-09-24   
558207            1068.0                  7295  2008-12-31  2009-01-02   
558208            1068.0                 4

### Cleaning Beneficiery Details

In [59]:
beneficiary_data

BeneID         DOB  DOD  Gender  Race RenalDiseaseIndicator  \
0        BENE11001  1943-01-01  NaN       1     1                     0   
1        BENE11002  1936-09-01  NaN       2     1                     0   
2        BENE11003  1936-08-01  NaN       1     1                     0   
3        BENE11004  1922-07-01  NaN       1     1                     0   
4        BENE11005  1935-09-01  NaN       1     1                     0   
...            ...         ...  ...     ...   ...                   ...   
138551  BENE159194  1939-07-01  NaN       1     1                     0   
138552  BENE159195  1938-12-01  NaN       2     1                     0   
138553  BENE159196  1916-06-01  NaN       2     1                     0   
138554  BENE159197  1930-01-01  NaN       1     1                     0   
138555  BENE159198  1952-04-01  NaN       2     1                     0   

        State  County  NoOfMonths_PartACov  NoOfMonths_PartBCov  ...  \
0          39     230                   12                   12  ...   
1          39     280                   12                   12  ...   
2          52     590                   12                   12  ...   
3          39     270                   12                   12  ...   
4          24     680                   12                   12  ...   
...       ...     ...                  ...                  ...  ...   
138551     39     140                   12                   12  ...   
138552     49     530                   12                   12  ...   
138553      6     150                   12                   12  ...   
138554     16     560                   12                   12  ...   
138555     21      20                   12                   12  ...   

        ChronicCond_Depression  ChronicCond_Diabetes  \
0                            1                     1   
1                            2                     2   
2                            2                     2   
3                            2                     1   
4                            2                     1   
...                        ...                   ...   
138551                       2                     2   
138552                       2                     1   
138553                       1                     1   
138554                       2                     2   
138555                       1                     1   

        ChronicCond_IschemicHeart  ChronicCond_Osteoporasis  \
0                               1                         2   
1                               2                         2   
2                               1                         2   
3                               1                         1   
4                               2                         2   
...                           ...                       ...   
138551                          2                         2   
138552                          2                         2   
138553                          1                         2   
138554                          1                         2   
138555                          2                         2   

        ChronicCond_rheumatoidarthritis  ChronicCond_stroke  \
0                                     1                   1   
1                                     2                   2   
2                                     2                   2   
3                                     1                   2   
4                                     2                   2   
...                                 ...                 ...   
138551                                2                   2   
138552                                2                   2   
138553                                2                   2   
138554                                2                   2   
138555                                1                   2   

        IPAnnualReimbursementAmt  IPAnnualDeductibleAmt  \


In [65]:
beneficiary_data['DOB'] =  pd.to_datetime(beneficiary_data['DOB'], format='%Y-%m-%d')
beneficiary_data['DOD'] =  pd.to_datetime(beneficiary_data['DOD'], format='%Y-%m-%d')
beneficiary_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 138556 entries, 0 to 138555
Data columns (total 25 columns):
 #   Column                           Non-Null Count   Dtype         
---  ------                           --------------   -----         
 0   BeneID                           138556 non-null  object        
 1   DOB                              138556 non-null  datetime64[ns]
 2   DOD                              1421 non-null    datetime64[ns]
 3   Gender                           138556 non-null  int64         
 4   Race                             138556 non-null  int64         
 5   RenalDiseaseIndicator            138556 non-null  object        
 6   State                            138556 non-null  int64         
 7   County                           138556 non-null  int64         
 8   NoOfMonths_PartACov              138556 non-null  int64         
 9   NoOfMonths_PartBCov              138556 non-null  int64         
 10  ChronicCond_Alzheimer            138556 non-